{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# SESSION_4"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Prerequisites: In a terminal, You need to create, install biopython and activate the `Conda` env as follow before to start jupyter"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**We will create a new env called curso_4**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "!conda create -y --name curso_4"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "!conda install -y -n curso_4 -c bioconda -c conda-forge mummer gepard racon nanopolish minimap2 jupyter"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Problem with Nanopolish"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**for MacOs users: you may have problem with mummer v.3 installed using conda**.  If you have troubles, you should install mummer4 using a regulation installation, see here :  (https://mummer4.github.io/install/install.html)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "!conda activate curso_4"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "!jupyter notebook &"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Polishing Assemblies"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "What is “genome polishing”?\n",
    "\n",
    "“Genome polishing,” sometimes referred to as “genome finishing,” is a workflow in which assembly software searches for local misassemblies and other inconsistencies in a draft genome assembly and then corrects them. Genome polishing can be used to create hybrid assemblies with Illumina data and long read sequencing data and is especially valuable for enhancing assembly results where there are concerns about single molecule or nanopore sequencing accuracy. (https://www.dnastar.com/blog/dnastar-news/genome-polishing-benchmarks/)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/jpeg": "/9j/4AAQSkZJRgABAQAAAQABAAD/2wCEABALDA4MChAODQ4SERATGCgaGBYWGDEjJR0oOjM9PDkzODdASFxOQERXRTc4UG1RV19iZ2hnPk1xeXBkeFxlZ2MBERISGBUYLxoaL2NCOEJjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY//AABEIAWgB4AMBIgACEQEDEQH/xAAbAAEAAgMBAQAAAAAAAAAAAAAAAQIDBAYFB//EAEsQAAEDAgQBBwoDBQYFAwUBAAEAAgMEEQUSITFBExQiUVRhkgYVFlJTcYGRodEXMrEjQkNEkzNicnPB8CQ0gsLhBzbxNWNkhLIl/8QAGQEBAQEBAQEAAAAAAAAAAAAAAAECAwQF/8QAJhEBAAICAgICAwEAAwEAAAAAAAECERIDEyFRBBQxQVJxIoHBMv/aAAwDAQACEQMRAD8A+foiICIiAiIgIiICIiAiIgIiICIiAiIgIiICIiAiIgIiICIiAiIgIiICIiAiIgIiICIiAiIgIiICIiAiIgIiICIiAiIgIiICIiAiIgIiICIiAiIgIiICIiAiIgIiICIiAiIgIiICIiAiIgIiICIiAiIgIiICIiAiIgIiICIiAiIgIiICIiAiIgIiICIiAiIgIiICIiAiIgIs/NX+s36pzZ/rN+q1pb0ztDAiz82d6zfqnNnes36prJtDAiz81f6zfqnNX+s36ppY2hgRZ+av9Zv1Tmr/AFm/VNLejaGBFn5s/wBZv1Tmz/Wb9U0sbQwIs/Nn+s36pzZ3rN+qaybQwIs/Nn+s36pzV/rN+qaybQwIs/NX+s36qeaP9Zv1+yaW9G0NdFsc0f6zfqnNH+s36ppb0bQ10Wfmj/Wb9U5q/wBZv1TS3o2hgRZ+av8AWb9U5q/1m/VNLejaGBFn5q/1m/VOav8AWb9U0t6NoYEWfmz/AFm/VObO9Zv1TWTaGBFn5q/1m/VSKR5/eZ9fsmsm0NdFs8zk9Zn1+yjmb/WZ9fsmljaGui2OaP8AWb9fso5q/wBZv1TSxtDAiz81f6zfqnNn+s36ppY2hgRZ+bO9Zv1Tmr/Wb9U0sbQwItgUjz+8z6/ZTzOT1mfX7JrJtDWRbHNX+s36/ZRzV/rN+qaWNoYEWfmr/Wb9U5s/1m/VNZNoYEWfmz/Wb9U5s/1m/VNZNoYEWfmz/Wb9U5q/1m/VNZNoYEWxzR/rM+v2Tmb/AFmfX7JrJtDXRbHNH+sz6qOaP9Zv1TSxtDAiz81f6zfqnNX+s36ppY2hgRZ+av8AWb9VIpHn95v1TWTaGui2hQyEfnZ8z9lHMpPWZ9fsmljaGsi2DSPH7zPr9lHNn+s36prJtDAi2BSvP7zPqnNH+s36prJtDaKghXsll65ccPV8mcFjxutlglldEGR57tF76gf6rpfw+pe3TeELz/8A08H/APr1P+R/3Bd7NPFTszzSsjb1vcAF572mLYh0rWMeXJfh/S9um8IT8P6Xt03hCy+U1bM1sdbh+JfsbhhZE++upvounpnl9LE9x1LASfgltoiJyldZmYw5P8P6Xt03hCfh/S9um8IXTjEKIy8kKuAybZeUF1sucGtLtSAL6LM2tDUVrP4cf+H9L26bwhPw+pe3TeELp3NfUM5Rjzyb2W5M6fG/ArTw6Weug5ZrH05aDEM5vtxtxKm9l1h4n4fUvbpvCE/D6l7dN4QurEoIfHmJfG0Zja2tl4GGYtJHVmOplc6N5tdx/KV0pF7xMx+nHk5Kcdoi37af4fUvbpvCEH/p/S9um8IXvY7PJBQ5onlji4C7d1rtjrTgvK87dntyua5vlteytYtNYnLNuSsWmsVziMvK9AKbt03gCegFN26bwhblBFiNdAZWVz2gG1i4rawmsqJ3T0s7ryMBs/6LVqWrnFvwzTmraYzXGfw8n0Apu3TeEJ6AU3bpvCFtRtrzifNDWvu3UnMbFdMNlnk2pjy6cVo5M/8AHGHH+gFN26bwhPw/pe3TeELsUXPe3t20hxv4f0vbpvCE9AKXt03hC7JQpvb2aQ4/8P6Xt03hCj8P6Xt03hC7JE3t7NIcb+H1L26bwhPw+pe3TeELsVKb2NYcb+H9L26bwhSPICmH89N4QuxUJvY0hyHoFTdum8IQ+QNMf56bwhdeib2NIcf6AU3bpvAE/D+m7dN4QuxRN7GkOO/D+l7dN4Qo/D+l7dN4QuyUJvb2aQ478PqXt03hCfh/S9um8IXYom9vZrDkB5A0w/npvCE9Aabt03hC7BQm1jWHH/h/Tdum8IUfh/S9um8IXYom9vZrDjvw+pe3TeEJ+H1L26bwBdiib29msOO/D+l7dN4Qn4f0vbpvCF2KJvb2aw478P6Xt03hCn8P6Xt03hC7FUD2k2DgT1XTexrDkh5A0w/npvCFHoDS9ul8IXWyBkjHNeRbjqsRoqfiz6lN7GsOX9AKXt03hCegFL26bwhdTzSDLlyab7rJGxkbcjNAO9N7ezSHJegFL26bwhPw/pe3TeALsETe3s0hx/4f0vbpvAEHkBSj+em8IXXucGtJJAA3JWk7GcNa/Ia6C/8AjCsWtP4SYrH5eAPISm7bN4Qo9Aqbts3hC6WaupYIWyy1EbY3Gwdm0KywTR1ETZYXh7HbOGxTaxrVyZ8gKU/z03hC163yGpqWinnbWzOMbC4AtGtguvqa+kpHBtRURxEi4D3WuseLEOwarLTcGFxBHuTaxrV8gCu1l0a1Z2CwXZyYbJZWAVsq6K6X/wBPRbFqn/I/7gvQrITjXlW6jne4U8A/KD1D7laPkCLYrU/5H/cF6uLYfX0eMjFMOj5bN+dn0OnUudZiLz7wXjNYaHlRgtPh0EU1IHMY52VzC4kXtof1W95TVstPgtJDC4s5YAOI3sBstbFocbxmna99FyUcR6MQPScTx1W7V4bWYxgMLZ4hBVwnosOxG31Wsxiu85w54/8ArWHizR4acPMcGG1zakDSUtO/frsukweeaXAIjUmZsjDluAc2h0WjHiPlJCxsbsNY8tFsxG/yKyw4rjxmaJ8KAiJs4sHSA6xqpfNq4/8AVpis5/8AG3UYiWMnpKfMa2xMbGtvvxXn4M7GYaSpPJggatExIN+Nl7TafJPG5jHOk5M3lcNTrsVmfMJqJ0kLXPzCwGxXmz4w99eTFJpiPP7a1HJI9reWc5spiLnsDeiTxN149BRCtpqpg/tGuuw969WrlrKZzmUkDpg8lxLho2/AKmA0s9OJnTxlmYi113pM0pNsvm8sRy8sVmPHl5k9aZ8J5vNflonga8QvaZ/7f/8A1/8AtWljeFvlkFRTMLnHRzR+q9BkMhwYQ5bSchly99lu9qzWsx7c+OnJF7Rb08bCocRkpXGjnZHHm1B3v8lsYE4QVk1NM0ioOpde91ioXYpQwmKOjDgTfpf/ACtrD6GqbUyV1U0CUg5Wd63yTExbMx5c+GsxNMROY/OfwpF/7ok/wn9AvdXNRtxIYlzs0nSdoRwH1XSjZcOaPx/j1/FnMW8ftKIi4PUIiICIiAiIgIiICIiAiIgIiICIiAiIgIiICIiAiIgw1UbpaaSON+R7mkB3UudoMExGHEBJJVOyMc0+8AWsuoUIPGijMtXizcubNLHYXt+6FMkRysHJb7gzG46uK2oKVvOawv6TZXtNuqwWdtHTtvaJovvos1naMw1aIifDzxFZpzQO01/t+C2I6Snkia92dhcL25Q6LPzGm0/Yt0UOoKZwI5IC/UtMhoYT69urOVsAWAHUpAsLDYIg57yzmfHhTGMcQJJLOtxFlzg83RYHGJaV76ubMWyA6CxsOK6nynoZK+khiicxrhJmu82GynBoTQYXHTvfA6ZriSc1xa9916K3itP+3mtxzbkn/HKSU9VBgIdOHMifMCxrvcbldB5MUoZTwVXP3EZTeEu0avSq3R1kToaiOnfGeDpF5sWCUDHuDYgA8FhPL8Ck8sWrifBHDNbZjy83y1ex9dBkc11ozexvbVdNW/8At6Yf/jH/APlebFguG00xLYGOLSbcpIT9F62Km+D1X+S7b3LF7RMRWP06UpMTNp/b5OwLIVAFgpAuujIApyqQpC2zl0vkILYpUf5P/cF18rKsvdycjWt4L55g+KSYRUPmijbIXsykOPfdesfLaqH8rF8yuN6TM+HWtow6vk60XvKw32023/8AHyUmOru39q21hm03N1yJ8uKvskXiKg+XVX2SHxFY67NbQ67kqwH+3bbvCNjrAHZpmG4NiBsVx/p3V9kh8RUenlZ2SHxFOuy5dlydZmP7ZtrEDTj1qOSrbAcsz5LjvTyr7JD4inp5Wdjh8RTSxtDseSrcv9uy/u2Tk62xHKs99lxvp7Wdjh8RT0+rOxw+Ip12TeHaCOqDtZARc/K2iryVb7dndouM9PqzscPiKn0+rOxw+Ip12Ters+TrCD+1aDwt8e73II6u3Slbtw6+vZcX6fVnY4fEU9PqzscPiKddjerszFWZujM23u71uDZcB6fVnY4fEU9PqzscPiKaWN6voCL5/wCn1Z2OHxFPT6s7HD4inXY3q+gIvn/p9Wdjh8RT0+rOxw+Ip12N6voCL5/6fVnY4fEU9PqzscPiKddjer6Ai+f+n1Z2OHxFT6e1nY4fEU67G9XfIuC9PazscPiKg+X1Z2OHxFOuxvV3yLgPT6s7HD4inp9Wdjh8RTrsb1d+i4H09rOxw+IqfTys7HD4imll3h3qLgfT2r7HD4inp9Wdjh8RTSyb1d+i4D0+rOxw+Ip6fVnY4fEU67HZV36L5/6fVnY4fEVPp7Wdjh8RTrsdlXfouA9PazscPiKentX2OHxFOux2Vd+i4IeXdXb/AJSHxFVPl7V30o4fEU67HZV36hcD6e1nY4fEU9Paw/ycPiKddjerv1C4I+XlYP5OHxFPT2s7HD4inXY3q7trQ0kgau3VlwXp7V9jh8RUentZ2OHxFI4pj8QdlXfIuB9PavscPiKn08q+xw+Ip12OyrvUXCendX2SHxFUPl7Vj+Th8RTrsdlXfEA7qMrfVHyXA+n1Z2OHxFWb5e1h/k4fEU0sdlXdiNgNwxuvcpyt6h8lxHpvWdkh8RVD5dVY/lIvEU0su8O6yt9UfJauLf8A0mr/AMp36LkWeXFY7+Ui8RUVPldU1NPJC6miAkaWkgnimkpvDmQ1XAspsoJXZzUCtdUupuurjla6q5yguWNzlMLEpJVSVUuV4opZs3JRPkyi7sjSbDvUl0iVCoVmte4Eta4hupIF7KFlvKERSgiyjKr2UWVZlSyhZLKpCrCqKbJZERZFayK4TKtkVlBUEIiIooUogBWAQBWQVKqVcqtkwZQgUgKwCkkIAQlSVVRqZwhWDdEaFkuLLUQxMsRCKXFVQFOyIgKQFFlOyAVFlZvejrE6IKLIxvFQG6rK4ZWIsMZ3UFEVRVLXVlbLZBUBWDbqzGi/SVnkNHRO/BBiebaLGr5SSrthPHRQYQ2624Iha5VWhg21Kztdlsoqz2gNWpkLnrYkcSbBQ1th3oJa2wsp2UhCo0qTZYnOUyOssIKGWQFCVVF1cglUKkrdwmGsNSKqjpjOaYh7ha4+KkzhYjJg0uHwV98WgfJDlIygbHvC9zAS+Gtm80vZLR1XRcwj9pDva4PVfvXleUtQ6uqoK/K1raiIGw4EaEfNYfJ+SqhxinkpIzJIHWLbX6J0K5TGYy6ROJw9yhp8QwTDZqY08fK1Ny98pAZE21hc8TvovLroMHhwiMU9QZcQDrPLL5T178FteW76qTGSJ2ObCwBsRtoRuT714uHU7qmuiYAC0ODn5tg0aknuspEeMrNsThrqQvW8oamGuqGVNHRGCnAyB+SweQvIW48pM4dRD5Jtlips2JRslqYw9kZZqdL9a82pwg0+H8uZHOmE5hMbWG2l9j8F1dJj2HspaEHE44xFC1ssZiJJNhpe2i1aPFsO5OmipszyK5zxCGkuDTm1+q57WdPDlZsNrYGOfNSzMa0AkuYQAF6OHeTrK3DRWy18dMwvLOm3j77rosXffydxV7qqSZr5Bk5RhZluR0RfdeThs+GT+TQoK2uFO7lS89Ek2v7k2mYTERLya7Aa6krjSiIzuy5muiBIc3rVsMwCrr6ySmc11O+NmY52H4D4ro245hk9dIBOYmRQCOF8gdlcb63A16ln8+4acVp5xWgMdTOiecpFnA3F/qm1vSTWri2YXXSOkbHSTPMZLXWYdCmG0Dq/EoqIu5Jz3FpJG3wXV4fiWGUTHnzjykgqM73SB5zj+6BxXmuraNnltz1s7TSl+blADYdH7rUWmWZrDWxLyfjpGllPXx1VTnycgxvS7+K8tuHVr75aWY2fk0afzdXvXV4zX0lc+IsxuMBk4e0CIgsHXe2q2nYnR4pi+GmkqtYZ3B0RaRymn59lmLWWa1ca3C6sSQienlijlkDA9zCBcmyy4lg0tHi78Pp89S9oBGVup06l1uI4xRUjayOaudUSPnGWEtN4rEX/AEVXYxhTMVq5Y8QDTWRBomaw/siPgm0+l1r+MuWoMGfUVU0FXytM+OMvsYiSVeswCqo8PpqsgvE4/IGG7DwBXTjHcNE8cbq3lTFSujdO5pHKONu7uWp5zoKnCMNbU4hI19O5vKxDNmdY737t1c2TFcOdGE1rJYmS00sfKuDWlzDuV6T/ACafDiZpKiZwbyZe2RkRIPcvbrsZw80FRFFWsMjZGSR5A69gRpc8dCrSYxhvnptX5xDo3QOZkLXWYdO7j/om1vSxWrimUFXLC6aKmlfE293hhIUxYZXTta6Kkme14u0hhIIXY0GO4eKeikNeadtNGWyUwYTnKpF5QUrGYa2KpEUYlc6ZgB6LTewOnuTe3pNa+3Elha4tcCCDYg8FBW7i0sc2K1csTg6N8rnNPWCVpLcRliZx4QVClFpnIiIgiyWUooIsimysG8UMoAsFCsVCqFuJUe5TwUgXRUsHFS82WS2Vug1H6rER1n4KkqkcRspylSDroFkcBcZLnREGZGfmFyquu5/csjIHbv0WQiOFoP5iisQjJGgTkmt1e74BQ+dzttB3LEdUGQyhujGql3PO6hrS42CzhoZoN1FhAbkHeqtdmkHvSR3BXp4iTmOyLMsuXW/FFaQ2CqNrlRYTZYZX20CyPfbTitR5JKYSZVJuiBCmGcsihSi6Mqlex5PSYtG+fzXHyjbAysNrELyCFt4bWVtLNydDO6J0xDDa2qxaMw1WcS9aTDue4TLzON7uTfyrIbEviJ0cz3bEHuXueRGHyUdLLJUwGKWV3Rzizi0d2+6xUNVVYDXwUmITmqmqiALHoxi/WdSbrwMeqqp2JjEI55BHLrEWuIyW0Le4grjibeHbMV8uz8qTh3my2JF2QvGXJbNfuXmR0mCNwB0tC1h5RuUmR+QuN/yuPAbLi6mrqKt4dUzPlcNAXm9l7mGMdLBhtFa7KtszHj3kWPwtdNMR+U3zP4Y8YqcUbhEVJU0DKWkEnQLW6aX0v/u68BetjE2I04GEVk7XspSMoaNNtNd9ivKsutY8Odp8pfFJHblGOZmFxmFrhTFI+GQSRPcx7TcOabELuZsJpa00/PXzPZHQCQWd+X3Lwcfw6hp6GhrKBsjI6lpOWQ3KzF4nws0x5edW4hiFU1ra2ed7d2h5NvfZYOQmzNbyMmZ+rRlNz7l0fldTTzebnQwSPaKYXLWkgL3aGGJmI0TpHyPkFCCzNYhu1/8ARTfEZaiszL589ro3Fj2ua4bhwsQgXYU+E4fiETsSq5ZJ+cTFrXSPyEAacNzpsvLo6enw/wArI4o2mrhY4kBoufynh3f6KxeJSazDxXsey2djm36xZQ1jnuDWNLnHYAXJXeYvQS4qymAqQaV84Dg+LI9vcvO5lh9PjFPDS01bDLHOGcq78h+KkcmVmjknxuY4te0tcNwRYhIpJaeVssL3RyNN2uabELuH4FRSzVVXWvdJmnLBeTLbh8StUYFhEEFXNMZZo4pg1pjdrY20PzTsg65cfK+SaR0kry97jcuO5KgBdpL5NYdT1VZM9s0lPDG1zYWHpEkdawUGB4bVSVVQYqiKCEC0U7shueJPUrvDOkuUAVw1evj+HUtDVRijkzRyMzFt75T715zWLrE5jLnMYnDFlVS1bXJqDGg1g1Cszm2WMhMGWIqqykKhCTBlVTZWDUIUwZUSyvZMqGVEsr2HvUjRTBkjic7ZWfYaDgjHOudeCoRcqihvdSAsgbpr81YRG+jSe9DLGGFxHAdazlkcerTmIWaGAWDni56lmMbD+6EV50jiTl6t7I2F7zo3TrK2ntigP5buOyxOne/bojuQTyDIxeR3wCOnDW/s2rC43KtGzMD1K4TK7Q9zS956KwvOY3WWV1hlCwoIUtaXGwVmtLjYLKSIm2H5kWEG0YsNXFVd0G9ZKloygvduVVrXSvsP/hTC5WhhLzd1w3r61uNaBoNgpY0MaG3UtAtfrURgMZzlztlSR2XU78AtiV2Rq0XuLnEqrlVzruVXDVSpcOkiKAIG3VgFbZBBCKyiy0yqnFWIVVJG9hcjJMYpZK6oe1jXgmRxuRbUL0Wx09Zi9ThdPLytLO8vhk9R9r3HduF4Cs1zmODmOLXDYg2IWJrlqLYZsRoZcOrZKWYtL2cWnQr0MVqcPNFhww+SXl4WnPe4ynf53vsvKc5z3Fz3FzjuSbkqtldfabfnCXudI8ve5znHUlxuSq2VrILggg6haTLe894ha3ODbkuR2H5epa09bUVFNDTyyF0UIsxttlmdLWNA/wCIeXbZQ+5HwVRNXnZ9QeHFZxDWZbLfKPFWxCNtW7KBlAyjb5LEMbxFsrJBUnPHHybTYaN6li5auuBnqLnYapy9bYEyzAONgSSAVMQbT7ZKHGK+gY5lLUujY43I0IusMddVRVfOmTvE5Ny++qyvfXsc5pllu1waQHX1/wBhHS1zWhxkmAdtqVcQmZWq8axGs5Pl6p7uTOZttLHr0WSXygxObk+Uq3nkyHN0G61+XrbE8pPYbm50R09a1oLpZhfvKYj0u0+2eDHsSgllkjqnAynM/QWJ61Q4rWOiljdO5zZX5334nr+iq+SvYATLMbtDtCdB3qWS1ri4GeRmUXOdxCYg2lnZjeICqNSKlwlLQ0nTUDbRZI8YxBtU6pFU/lXDKT1jqstYzVzSbyym29iSAs0Ulc94aHym5tfWyuITafalTPPWzGWpkdI/a5UNZZbTZav15T1b6q/KVYFy6W3Xqh+WplCgtW7nq7E55ABvdUM1UP35dr8dlB50jViLV6fL1RaHNkkcDrpqofLVMIBdLci+l1qJR5ZYgj1XpNqKtxs0y8evgqOrKptrulF9r3QauUBtgNVj5Jx4LdNVX5S68oAte9+KxisrnC7XykdwKhiGtyDj1K3N+twWR1bWNJDppAeomyxS1M0wAkkc8DYEoeFuRYN3pkhG7rrFl70Hd80wmWwxsO4096rysbTZrLrCTcdyr7lcLlnNT1MF0FRI7RoCxNYXbbcSthseVugJA+qKyxvNgH723Cl0sbdzr1BYoQTme7dYCM7yb7ngmEmSR5mkvawGyrbg3ZZ3mOMENGY23KwalVMga0bm/cFsxzWZ0xZuwsFha3/4USG5twCuEyodSSjWFxsFZrC42CyuIiGVu6jUKkiJtm7qjG5jc7I1pe63HrWR+lmMCLlR15H5W6rbjY2JuXc8VEcfIt63n6KWAucLflHHrUkVDeUcSNuJV3OEYAGgHBXcRG3q4BakjiSSduChljleXm5Op+ixkaWVtyoK1hMq2V8ma3uUsbci+3FZnAOaCwWHUoZYHWGyra6uW233UFBCK1lNlpnKiiyvZRZQyrZLK9ksmEyqitZLIZVRSVCK2mVmQucIgXP/ADEnfrVX1b3NIAsCC3frt9lsfsJADI5oNuiwEW/0t8VjLaUuDRoLEl1+N9vksNZlVtYc5c5gcS7MNdtvsqPqXPJJGhLdL7WWbJTEtN25TbMcxuPcOpYp2sa8OitpvY31ViISZlkbXPAILGm7ibnq10+qjnhztdkuW347j/YWV74pS3MWk9Z0/dGnzuokFIHACzhte50Gv/hPB5YjWkuuWC4N22Oxtr71R9W58b2OAs4knrutjJSulDbA343O99AsU0TI5IS5rQCAXtab8T/oi5lDaoi12XLdW68bWUx1JEsshB/adRsRrdXNnE8o+J5/cF7D42WRsdK6SzcpbwOY3J137k8J5Y3Vpda8Y0uRY9d1IrXGTNl4WtfvumWmyEktz26Qa7QacP8AZWWNtKHlwyWDuiCT1j/S6vg8rMrDkcA2xeOkbrLzw5rkA6k2J61h/wCHyi5ZexzdLbqt1qpNO7NbK3UgEuO3WmIXyzur7EEM2Gmvv+6wOqxmcQz824LrgqIzE6iLXPa1+p1393/lTE5uWKz4wwWzA73vqriEzLEyqdGxjQNGuDt97En/AFTnRuLtBAN7X7gP9FkdHAWkXZe3RObUu7+5Rlp2A5TG7ouuS7W+XSyeDyg1mdxJjGvftpwVed9IuMdze4120spa5vJxtzsbHYB443vqrsbS5mh4ZmJs4NcbAa6/ongzLAKqz3OEf5jc3cd73UisyhrWRgNHC977/dZzHTGIPIYAdHWcdNOHxUNZAxj7OjOZ2muzbi3x3WcQuZaUzzLIXHTYWVPot7JTOe5rGgkfl1PS04/FYaiFrJy1ug00+CsMy1xcmwViNLBZi9jIw0N1I3WIO14qogtNhfT3qzIweJt3cVk5ME5rWHUqufwb8SjX+pLg3otAuPosfKv1AcbFFeBl5R3aphM5ZJBycAbxOiwAWIHErNUO6Y7ljY0m5tw3VgmVSL/EoBZWPUPmrxMa53S24ImVPytJ4lUawvNgsrmF0haOHFWcRGMrd0WFXERNyt3WGxJ71ZZY22Gd3wV/C5yi3Js/vFZYo+TGZ35z9FkZG0ASOHS7+CWzO1+SyIF3fFX0Y25NlYANBJWvUSscwsbqVMZMsbi6SY32bwWOTU2GwWbKRFnA1cLLBY9S0mUAWCBvE/LrWURmwLtArMDLOJOo6uKClrDpbn90I5125R71BI/Nrcps4ADVFQx2UG4v1Kh1KyP3sBoqFRMrWSysi3hzyrZRZXRMGVLKbKyWTArZQVZVUVWyWVrIphcq2SytZSAmEyrZFayIKqLK1kQQNENybkknvUoi5VsisiCAFkAsEa1HGykrCHFVRETIpBUWKsArCZDqgYQLkWVm2BFt1aaUyWFrAKymWHT3pdTZTawUVW2n+iW61YdyW1txQQ3Q3boVP5nWv7yoOmg+asxpym250QUPSdp8FmZGGDM61/0UtaGC/wBVRzi83OjeAUa/A599tG8e9UI6tlbfdS0XNuBVTOVLWGi2adoZGXlY3R2dlBB7ws09mRBiGWq4lzi48VdrnBmQbHUoyMyOs25WQZG5rgk7BVMsYapALjYbKWNLlZ7gwZW7qrCrnCMZW7rCVZSG5iAEwZyRszHuC22Rj87thsFaKMNbqNEkOtm6lZlVXG7rDc8OpXawNGvxKljAwXO/ErXmlvoNv1TBlWaXO6w/K1YALnXZXtZoHE6qbWNurUrWEVedbA6BZICRdztRsAVja0udZZL7AbX0UMsbnOe65PwT94AbDdSBYX48FBFhlG/FBD/zabcEA3PwCyMjLrX0skmRpyt6VkMsdrt04KuVXB+Si2qgmyLPzOp7NN/TKGjqezTeArWYY1lgRZuZ1XZpv6ZXu4DgHLWqK1hy36MThv3lZteKxlunHa04h49HhlZW6wQuLfWOg+a9WPyTqHD9pURtPUASukqa2lw9vJkB7wNI2bD7LzX+UFQT+zjiYOq11zjt5PNYxDd7cHF4tOZebJ5JVAH7OpjceogheTWYVWUNzPCQ31xqF1DMfqAf2kcTx1WsvSpa2lxBpYLMeRrG/Y/dS0ctPNo8FLcHL4rOJfOLKbLpse8n+SvUUUZt+/EBt3heDzOq7NN/TK3W8WjLN6WrOJYLIs/M6ns039MqeZ1PZpv6ZW8wxiWuoWwaOq7NN/TKjmdV2ab+mVMwYlgsostjmdV2ab+mU5nVdmm/plMwuJa9kWxzOq7NN/TKnmNT2ebwFMweWsArBq2OZVPZ5vAVZtHUj+Xm/plJmCIlgtYLGRcrZdS1R/lpv6ZVeZ1XZpv6ZUiYWcsFgnuCz8zquzTf0ynM6rs039Mq5hMSwe9QtjmVV2ab+mVdlFUBwzU01v8ALKZgxLXaCG361azQw3/Mtmanndbk6SZoH/2ysQo6q/8Ay039MpmExLAT1IR1rY5nU3/5ab+mVHM6on/lpv6ZTMGJYLXUn8unxWc0dTsKab+mVMdDUnV1PNb/AAFTMLiWuyMvPd1rM8iMW6th1rOaWoZe1NLcbWjOixCjqibmnm8BU8L5hgNzq7ioWc0lUT/y039MqRRVPGnl/plazCYlrgK8bC82Gnes3NKns03xjKsKWpAP/DzX/wABTMHlEcYD7Xvl3WOc3k0G2i3IaSdrCeQluf7hWIUdQf5eXwFMw1iWtcgaEqWxl9hw3K2BRVLnf2Eo/wCgq8lNUMGRlPL78hVzCay1pHhgys3WCy2OZ1PZ5vAU5nU9mm/plXMGJYLLcghDWhztyrU9BOTmfBJYbAsKzvgqDo2CXwFSZgxLEbk6DZGsyk8Sdys/N5wAOQk8BWKeGpJyMgl13OQqZgxLXqH3sxp04la1szgOC2pKSpvYU039MqGUdTqTTzeAqxMGJa506XwCi3R96zupKon/AJab+mVZtFUEi9PLYf3CrmExLBbKz+85XYy8gB0AFlnbSVAfnMEpP+ApzeqdmJgmudAMhUzBiWs4NBu0FwGyr0z3e5Z3UtV2ebwFQKOqGvNpr/5ZTMGJY2NN8o3I1WItykg7hbTaKqDSebyj/oKGkqOzTE/5ZTMGJaoaSrloyg314rLzSqO9PN/TKnmlSTbm039MqZgxL6IiIvnPqoWtidZzGl6H9tJo09Q61tAXIHWuex2UyYnIODAGj/fxXf49Ivfy8vzOWePjzH5l57iXOLnEknclQrMYZHtY0XJNhZdhS4bTQQBnJNcbaucLkr38vNHFh8jg+NbnmcS41S0lrg5pII2IXSVeARSyPkhfyYI0bbS65twyuIPDRa4+WvJHhjm4L8M/8nT4ZWc+pen/AG0eju8da2lz+AymPEmN4PBaf9/BdARYkL53yKRS/h9r4nLPLx5n8waqF5HP3xRyzPmL5m5r0uUdGxsO8e/itWpxSqcc4eIcrCWhrm2eQ62g1J9wK4PU6JQvFnqKgBszZZ7ZpbxNy6ht7bjjZXjxaVxYHRwxhxPTc7outbQWvrr9EHrqV42KV1XSy1TohmhZADo25a43s7vGiscWlM7oo4mOOYNaSbfvZTf9UHrovHbi8uZrXQsBuAW3N3G5BLe4WuVWLGpJI2HLAA8NJkzHIy4Js49en1Qe0i8Q4zUcmJOata0uDembWOXNrt7lOKVtRGKcsk5uXwOkLS5o6QtYag334IPaS68fzvOZJGNp2nID+dwadANTc7G6vS18lRLUdNmUQh7GgEG+tzr7kHq3ReLSVdeyOF0zS5kwaQZSMw6Jc4jLw2sN1MeMvfcFkLOlblHu6Fstxtx4fBB7KLxTiU8UsufK8skcA0aBo6Ni7u1JWRuKyuq44RFG8Oy3ex2jrk6tv1W70HrIvFra+ppKzIS90bSXus2/QNrfHN0fcUOKVNO10cjGSTMBBFyDcNvmP92+iD2kXjtxKpjlqo3xtLmElpcbNdo3QHuvxVY8UnL3TCMSQktblFwb2Ow94Qe0i8XzzJ0Ran1/iZjk/Le3v4f7svWp5DNTxylpYXtDi07i42QZFClEBERBClEQEREBERBClEQQpREEIpRAUKUQEREBQpRBCKUQFClEEKURAREQG6OB71zmNMLMUmvxII+S6NebjtIZ4G1UYu5gyvHd1r0fGvFb+f28fzeOb8Xj9NTyeMYrzyls2XoX611S4EEtNwSCOpbHP6vtEniK9XN8eeS20S8HxvmV4aazDtlymPxwx137IAEtu4DrWp5wq+0SeJYHOc9xc4kk7kpw/Htx2zMnyfl15qaxDdwRhfikNv3SSfkukdq4+9ebgVIYIXVUgs6QZWDu616C8vybxa/j9Pf8LjmnF5/Zp3KMrdNAubpqmooMQxCqkeX0LqpzJB7I6WcO7XVZMMrZBhNKyGdvKSOlcLxulJAkOtgdu9ed7HQ2HUlh1BeFBi9bWvoo6aOFj6iKRzjJezS1wGg4+5Wgxarqm01PE2FtVKZQ97gSwBhsSBubkoPbslh1BeIMVrXzR0rWQCo5d0MjjfLo24I+yl2K1cXKUr2QurBM2JrhcMOYXueI46IPZLQQQQOpQyNkcYja0BoFgF4lTitbRmWnlZDJUMdFlc24a5r3ZdRwKtNilY2rkpByYlhaHSObA+RridQBbYW4n5IPbsOpCAdwCtE4lyeCmvnidEWx5jG4ag9XzXl4HVuhNRTCR1RK+LnDQbi7/wB5ovwvb5oOisOpLDqXjQ4nVup5tI5alkeYU/JuY8G/UTqO8Kj8akigaBJHLNJKIxaF7Sy4v0malB7lu5LC2wXhPxisip5f2Ic9skbWSOidG14c6xFjqCPipfXYs2asp/8AhM9MwS57Os4EE5bdeh1Qe5pfhdRpcbXXgQTVlVi80lJyMZkpInnlbkAm+gA/VTT1c1ZXRTxta2fmcgynVucOt8roPfsOpDa/C682jxJ1ZNTNiaAHQmSa+7De2X33v8l51a2GTG61s9JVVBEbMnIk9E2PURZB0dh1KOiOpeHR1GJuaKOMwialhY6V013FxN7DTuGpVIKznlZzrJYPw57iwnS4fYj6IPdMTHOY8tBLL5T1XVwvBbiM8MFLYR09MYGO5R8bnMJO7bg9G2mp61hFVWU0mMVNPyJhhmD3Z7kuGRtwOr3oOkupXPz4jLBU1TqaCMyumgj6RIzZm8fcskmKVlLzqGZsMk7HRtjLLtac5sL36kHuIvNpqmtbiYo6vkHDkTIHxgi5vbY7LXxTFZqM1D4XMeymAL2CJzjtexcNAUHtIvDqcTr/APj307acR0gDune7hlvbTb3q5r6ysmnjo2wtZDE1z+Vvd5c29hbb3oPYGqLysGnbT+TFLUSnosgzOK0MErSyrdGZHSyVcRnym/RkG7Rfut8ig6VF4lNidY+CXMI3VTY84puTcx7T1anpDvCq/GpIKNznPjmmMjIwwROYWl3rN1PDhug91F4YxOvEczW07pnsDXNkED2CxNj0TqSBrpurHEqk0gfC6OYiTLI9kLrxC17uZe90HtIvGGI1U0lJDSyU0jp2vzSWdlGXu3+CwwYriDoIqiVlOIxU83ka293HNlzDq14IPfReA3G6mVz5YYi6JspYIhA8ucA6xOYaX7lM+K17IamqaynEFPNkLTfM9tx8jqg91SvIwrnJxXFBLKx0bZgA0A3HRHevXQEREBERAREQEREBERAREQFLXZT1jiFCIPLrsEbKTJRkAnUxnT5LyZKGqiNnwSD3NuuqVhI8D8xXpp8q9YxPl4eX4PHecx4cnHQ1UpsyCQnvbZevQ4I2JwlrSCRqIxr816pkef3iqpf5V7RiPBxfB46TmfKXOzHqHAKEULzPcwx0cMYnDWaTuLpAdQSd1rMwaijjhZE18QhDmsLJC0gE3IuOF1rxYtIx1WyoYMwle2mA/i2dbL772+amLFnQ4bDNVtYZCwukyvAtYkGw47INumwukpJI3wMLTGHBozEgBxuR8wqOwmkdC2PI9uV7pGua8hzS43NiFLsQby5pxG7lbZh1ZLfn93D3rXjxj/hhJzeWRrWRukeLD8wHD47IIqMFhc6jZEy0MMjnvOc5iSN773vxWw3CaNtK+n5NzmyOzuc55Li7gc291R+Iu5VsTonwyCUMLSQbgsLh+ipBiLhFE+U3caZkhBIaCSSN0GZuEUbY3MLXuLnte573lznFpuLk9SyVOHU9TNyr+UbIW5S6N5YXDqNt1gZiwkMbIoHSSPc9tmuFgW2vr1arK6rkmww1NLETIRowi5BBsdONtUGWWigmgjgcz9lGWua0HTTZKmjhqpI5JmkujDg0gkaEWIWpDiGUZTNzmVzg1kXJ8m8G3EHhode5TJXVDauGMUr7uje4x5huCNb9WqDI3CaUFxJme4tyBzpXEtF72BvonmmlMb2vEjy8hxe6Ql1xtY8LLC7GWiN0jKaR8bIWzOdcCzTf66K0mJWkyPjfC4OYdbG7XE/LYoMvmul5IxuD3gvbIXPeSS4G41WU0cLpZ5S0552Bj9dwL2/UrTdiM0jqZ0UDmxzBzmlxHTGUke7gtrDZ5anD4Jpmhsj2BxAN0GB+C0b3B1pWOEbY8zJHNOUbDRX810rH08kTHMdTNyxhjy0W6j1/FY34jbFhAHx8iCInAnpZyLj4bD4qz8WhjmcyQANaHHM14d+Xe4GyCuD0L6U1M80bY5amUvLGuzBo4C/zPxW4ynjZUyTtB5SUAON+rZaVVX1DKYSCmkjJkjA1DswLgLdxVnYsyO7ZY+TlEnJ5HyAC+XNfNtayDJV4ZTVcvKyCRr8uUmN5ZmHUbbhHYVSOkjeGOYY4zE0McWjKeBHFYm4uJHBkFO6V2VziGuFhY2Ov6LLNWE0tNPDo2Z8Y1HByCr8JpXxtjPKhgYIy1srgHNHAjirnDaUw1MOQhlT/AGgB30A+GgCx19e6iqW3YZI+TByt3JLw0fqodiDjK2MwvjkEuQsuDm6BcNfggyuwylc9zyw5nPY89I7tFmqlbh7Jo6hzIo3yzBocJCcpA22296rJi8Eccb7HK9gdckCxJsAb/H5FQMXY8sZHFykjpOTyseHC+Uuvfq0QYsMwySnrXVUrWsPJ8m1oldITre5cVsz4RSVD5nSNfab+0YHkNdpa5CxyYmHQ1JawsMLXfvDMCP7v6JBiEuaUzxWjbUCIODtgQLE/Ej5oM4w+nEU8eV2WoAbJ0jqLW/RUlwijmka9zXtcGCM5JC3O0bB1t1sU84na9zQQ1ry0E/vW3Pzv8lrc9qOeGmEEWYtcW2kvlttm00ugv5tpeYChyHm42ZmO1729yy1NJDVmMzNJMTszCDaxtb9CtWPEJWtrH1MUYjphq6NxOY2uRqOCq/EZqdwZUwM5V7QYwx+hJNrEkaWuNUGVuE0rSSeWeS3IC6VxLR1A30TzTSFkrZGvlMtg50jy52m1jwsslPVPlM0b4g2eG12NfcG401stYYnNeSMwRmUOaxpZJmYXG+hNtxbVBmbhlOI3svM4vIJeZXZtNtbqBhVMGkAzB5dnMgldnJtbU+5YpMSmhhlL6bPJDI1jwx3RsbdK5147LJLiDo3yuEAMELgyR5fYgm2wtruOKDJBh1NTvifG12aPNYlxJObcnrUDDaYQciGHJyvLWzH8181/mtxEGkcKpeXMoa8XfnLA8hhd15dlZ+HUz6eaBzCY5nl7xfc/7C20QarKGCOtkq2ZxJJ+YB5ynvttdbSIgIiICIiAiIgIiICIiAi83z7h/tj4Cnn3D/bHwH7LfXf059vH/UPSReb5+w72x8B+yefsO9sfAfsnXf0dvH/UPSReb59w/wBsfAfsnn3DvbHwH7J139Hbx/1D0kXm+fsP9sfAfsnn3D/bHwH7J139Hbx/1D0kXm+fcO9sfAfsnn7D/bHwH7J139Hbx/1Dajo4YzfIHHlHSgu1LXO3t81gOEwFhYHyta5hY8NdbMLk6/M7Knn3D/bHwH7J59w72x8B+ydd/R28f9QzDDoQ8SXeZds5OtrWy+7uUMw2BlO6BubI5rGnXWzQAP0WLz9h3tj4D9lIxugP8Y+Appb0dtPcM8lDDJUcs7NmzB2+lw0t/QlYW4TA1sYEkuaNrWsdm1aASR+qqcdw8fxj4Cnn7DvbHwH7J139Hbx/1DPDh8UMwlDpHPDnOu517lwAP6BXbSRspebtc9rbkgh1iCTff3lavn3DvbHwH7J5+w72x8B+ydd/R28f9QyHC4XPMj5JnzaWlLuk217W+Z+ayx0bI5GyZ5Hva1zcz3XJuQT+i1vP2He2PgP2Tz9h3tj4D9lOu/o7eP8AqGVuGQNgkhGbLJEIjr+6L2/Uq8tDDNPyr75rNG+nRJI/UrX8+4f7Y+A/ZT57oPbHwFXrt6O3j/qF48MijfERJKWxX5Nhfo0EWsstJSCka1jJZHRsblax5vbqWuccw8WvMfAVHn3D/bHwFNLejtp7hnOHU5gdEW3zOzl5/Ne97396xjCoOi1zpHxszBrHO0AcCCPqqefcPtflj4Cnn3D/AGx8BTrv6O3j/qGYYfHlDXyzSAFpbnfe2U3CmTD4ZJHSXc2Qvzh4OrTbLp8FgGO4f7Y+AoMdw8/xj4CnXf0dvH/UNmOjjjeHguLshYSTuCblOZQmibSEExNaANdRbY361refcP8AbHwFXbjFE4XEpt/gKddvR209wscNicHcq+WV7i273uubNNwPmsjqKJ1Ry5zZ84fvpcAj9Ctc43QDeY+Ap58w+2sx8BTrt6O2n9QyDDKYRyMs60j+UvfVpvcW6tf1V2UMbXxvL3vfG8vDnHckW/QrAMboCLiY+AqwxiiIvypt/hKmlvR209rS4bFOXGWSV4LXNAJ/KDvb/wAqtRQnkKoQEudUD8j3WaDa19lXz5h/tj4Cnnug9sfAU0t6O2nts09PzcMa17ixsYaGnrHH3lYoKDmxkdDUS3eXOs6xGY8TpdQMWo7X5Q2/wlT50pPaHwlXS3o7ae4XhomR0rYHOL22IeDtITuT9ViOGsEUgD3SSOaGgznMA0G9v96qxxSkH8Q+EqDi1GP4p8JTS3o7ae1KbDBHHMHPyOm/NyRI0sQBc6ne91ePDWMp+QM0roxbKCQMhGxFgnnaj9qfCU87UftD4SmlvR209j8NjfTvh5WYco7PI/MMzzpvp3BYqjCzPKM5jcw5cziDnNvdoffwWUYtRe1PhKedqP2h8JTrt6O2ntuqVoHF6IbynwlQMZoT/GPgKmlvS9lPb0EWj52orX5U2/wlVOM0Nr8sfAVeu3pO2nt6CLzxjNCdpT4Cp870XtT8GlNLel7Ke2+i8841QDeY+EqPPdB7Y+Appb0nbT29FF53nugv/bHwFDjdAN5jp/cKddvR209w9FF53nzDx/GPgP2UefMP9sfAU67ejt4/6h6SLz/PVAP4x8BUee6D2x8BTrt6O3j/AKh6KLzTjuH+2PgP2Tz7h/tj4D9k67ejt4/6hyJUFWKhfUfDVRWsllBFksrWQqplSyKyWUVVFNlNkFUVkAuUBjblXPRCsBlCxONys/mW/wAQqlrqQFJ6gtMIOiqrWRBCKVkYzi4e4IKBl9ToFluAMp34dyhzsu2p+gWLW91F/A6+Y31KAX+KyAZ7C2+3cpkjMXRduepXCZYjrtsEViD7laNrL9IE9QCCljsN1ZrC7Ro04lZhAALuOncoMn7sYU/xf9Axkerjcqr5C7+61VJsddSqkkq4MhI4KWtzG52/VGNvqdv1V9XOyt3/AEUlYS0ZjYbDcj9Elf8AuN2Cs9wiYGN3WBTC5SFtQwWGaQfBRBBs5/wCzOkN8o36kMocdbAe5XvYC+pUNbkF93LHI/IOtxQS94b71Vlz0zt1lYwNbv3PBS+Q/lGgHBRVi7q+aX0WO9h3qTwCjXhdp4q19O9Yr21PwVmklUQ9Whj1u7TqV42B19VLuiDm6lYhJn9MMz9bbAcFYwlwB2G6htnnbbiVSWRznWuQBwRnIXt2A0HBDIbe/gqAXsFLuFkXKS0WuPkodsAFAdY6fFXc1pPQJA6yhliWQMzR5nae/inRbsQT1lWY1z+/rRGJxbfbMqlxH/hWc3KSOKgN4nZVmUbDMd1B2Qm5UlBG6iyK7dkEOGqhXcqLTIiKVUQoKsoUEIrAK2SwTC5Y7KVKWRMqrKxul0Yy6s8gC11mW49scjr6KgF1kYwOcMxsOtS9rQbMNwtRDMyxlQrHTgl/ghlWxUhp6lIDnGwWRrQ0b+8qSR5Q2MN3Iv8AQI54tZpOu561Vzs2g0CrbVFyXHAfNSLnuCWA3Tf3KosHhrCLZidieCh5Jsb7hGsc86D4rYDGRtF9SpkwwxwOdqdAsj3tiaRGAXWUPkc4WGgWPRu+qYXOGnFXS1VQImRjPwF7Dv8Ako86PmeImQtz7b/NbVhmuGgHrAUBjQbhoB67Lnrb27b8f8tOKv5WXJkt1knYcfkhq3GdosBGHlpsekbC+y3WQsOuRtttt1ZzYw64YzN1gbKa29rvx/nVosxJ8gaGQdJ5AbrpqFLMQfE4uyNLSxpLS6xBuQbda9CKCOJucsaOI0WJ0bHvB5NlxtZo0U1t7Xs4/wCWoKt3IZz0nukLbE2421SnxAxAZ4+UeS7Y6WBtYL1I6aKNl3xtJd+7lGqu2CPOC6NhN7gZRp3prb2Renn/AIsrmnJmbu7YHgjWiMdbjuspsG+5YcwzdLZdZcIHuLRpuVql9j0dXdameXO45fy9SoNPeVFyuDYEndVHWUO4aOCE8Bsi5S3Vyt3nikTeiXHZQ5wB01PWoZO9+ncrMvI6w0AWNoL3W4rYbZvRHAXJVwmVnvA2FrbLHqblx1txUXuf0HUpuLOHDr60VjcdgEtn14jfvUcboGuNiNO9VEN1uepTa7RbbrWV7WhgLdSdwsDnEhEyt0R3lY3OJOqKQLqplUalZM+TQH3lV02HzTIXO6Iuhlf8+/BVf1DZZMhaLW1KoRbimFyoBqhCyCyOsgxBqsBZERFnhUsvAPlBVH+HD4T91Hn6q9nD8j91x+xR6fp8robKFz/n6q9nD4T908/VXs4fCfur9mifT5XQWRc/59qfZw/I/dPPtT7OH5H7qfYofT5XQXS5XPefan2cPyP3Tz7U+zh+R+6fYofT5XQKWi5XPefan2cPhP3Vh5QVQ2ig+R+6n2KLHw+R0n5QsRNyufOP1R/hw/I/dR59qfZw/I/dI+RRZ+Jyy6HgoXgHHqn2cPyP3Uefan2cPyP3V+xRn6fK6G5UtaHd3eud8+1Ps4fkfurekFV7KD5H7p9iix8PkdJlDW93duVic4uK584/VE3yQ39x+6keUFUDrFA73tP3U76E/E5XQRhubp3t3KD3LwDj9Sf4UA/6T909IKr2cPhP3V+xRPp8r3w0k6BZ2QAC7yPcuab5RVbdooPfld90d5Q1bjrHD8j91PsUX6fI6UvH5YwozhrSMt3HiVzbfKKrabtjgB/wn7qpx+qP8OHwn7q/Y44PqcroSVVc/wCfan2cPhP3Tz7U+zh8J+6fYon0+V0Cs1t9Tsud8+1Ps4fkfurekFV7OD5H7qfYosfD5HROdbQb/ophjB6Ttgua8+1Ps4fkfurnyiqy3LyUAHc0/dO+i/U5HRSOMjrDbgs0UQiFyLuOwXLx+UNVGbiKC/e0/dSfKSsIP7ODX+6fup30PqcjqRd77DfiepZbZdt1ybfKasY3K2GnA/wu+6O8pa1wI5OAX6mn7p30X6nK6aSWxLW6kbngFhLjkc48dAua8/1VrcnD8j91Jx+qLQOThsO4/dO+ifU5XQN6zsrDi4/Bc55/qvZw/I/dT5/qrD9nBp3H7q99D6nI6HYd5QC++y53z9Vezh+R+6k+UFUf4UHyP3TvofU5XSyyFzWt2ACxDUrnzj9UTfk4fkfupZ5Q1TTcRQeE/dO+h9TkdMP2YsNXn6I55EYjABJ3K5n0hq9f2cGvcfunpDV+zg2tsfunfQ+pyOkvbQfEq4YTG4ggDrK5geUNUP4UHhP3VvSSr4xQH/pd91O+i/V5HTMiJGYDTrKxufqbagda58+U9aWZeTp7f4XfdYfP9Va3Jw/I/dXvon1OV0jiS0HjdRcOHS361znn+qtbk4fkfunn+q9lD8j9076H1OV0RaRuoJ4Bc+PKCqtbkoLe4/dR5+qvZQ/I/dPsUPp8romDVZY3iK/ElcyPKCqH8OH5H7qPP9Uf4cPyP3T7FD6nK6aVxdqsRXP+kFVb+zh+R+6jz9Vezh+R+6fYofU5HQhFz3n6q9nD8j908/VXs4fkfunfQ+pyuhRc95+qvZw/I/dPP1V7OH5H7p9ih9TleUiIvA+sIiICIiAiIgIiICIiAiIgIiICIiAiIgIiICIiAiIgIiICIiAiIgIiICIiAiIgIiICIiAiIgIiICIiAiIgIiICIiAiIgIiICIiAiIgIiICIiAiIgIiICIiAiIgIiICIiAiIgIiICIiAiIgIiICIiAiIgIiICIiAiIgIiICIiAiIgIiICIiAiIgIiICIiAiIgIiICIiAiIgIiICIiAiIgIiICIiAiIgIiICIiAiIgIiICIiAiIgIiICIiAiIgIiICIiAiIgIiICIiAiIgIiICIiAiIgIiICIiAiIgIiICIiD//Z\n",
      "text/html": [
       "\n",
       "        <iframe\n",
       "            width=\"560\"\n",
       "            height=\"315\"\n",
       "            src=\"https://www.youtube.com/embed/jCVkHq9dlGs\"\n",
       "            frameborder=\"0\"\n",
       "            allowfullscreen\n",
       "            \n",
       "        ></iframe>\n",
       "        "
      ],
      "text/plain": [
       "<IPython.lib.display.YouTubeVideo at 0x109717460>"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from IPython.display import YouTubeVideo\n",
    "\n",
    "YouTubeVideo('jCVkHq9dlGs', 560, 315)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# CALCULATING DIFFERENCES BETWEEN ASSEMBLIES"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In the previous session we assembled the b. subtilis genome without  error correction of reads. The accuracy of such assemblies equals to the base accuracy of sequenced reads and is unusable for many downstream analyses. \n",
    "Let us check the actual error with `dnadiff`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can now run `dnadiff` from the `mummer` package, which is a tool for calculating differences between two genome and providing a detailed summary. Try with the different assembly done at session_3.  \n",
    "(see http://mummer.sourceforge.net)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/bin/bash: /opt/anaconda3/envs/curso_4/bin/dnadiff: /usr/local/bin/perl: bad interpreter: No such file or directory\r\n"
     ]
    }
   ],
   "source": [
    "!dnadiff -h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!dnadiff -p bs_assembly_miniasm data/bacillus_subtilis/bs_ref.fasta data/bacillus_subtilis/bs_assembly_miniasm.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/guyot/Piracicaba_ESALQ_2023/Session_4/data/bacillus_subtilis/bs_ref.fasta /Users/guyot/Piracicaba_ESALQ_2023/Session_4/bs_assembly_miniasm.fasta\r\n",
      "NUCMER\r\n",
      "\r\n",
      "                               [REF]                [QRY]\r\n",
      "[Sequences]\r\n",
      "TotalSeqs                          1                    1\r\n",
      "AlignedSeqs              1(100.0000%)          1(100.0000%)\r\n",
      "UnalignedSeqs             0(0.0000%)           0(0.0000%)\r\n",
      "\r\n",
      "[Bases]\r\n",
      "TotalBases                   4045677              3931083\r\n",
      "AlignedBases       3845046(95.0409%)    3733242(94.9673%)\r\n",
      "UnalignedBases       200631(4.9591%)      197841(5.0327%)\r\n",
      "\r\n",
      "[Alignments]\r\n",
      "1-to-1                           323                  323\r\n",
      "TotalLength                  3847844              3731729\r\n",
      "AvgLength                 11912.8297           11553.3406\r\n",
      "AvgIdentity                  85.3145              85.3145\r\n",
      "\r\n",
      "M-to-M                           346                  346\r\n",
      "TotalLength                  3886262              3768634\r\n",
      "AvgLength                 11231.9711           10892.0058\r\n",
      "AvgIdentity                  85.3034              85.3034\r\n",
      "\r\n",
      "[Feature Estimates]\r\n",
      "Breakpoints                      690                  691\r\n",
      "Relocations                        3                    8\r\n",
      "Translocations                     0                    0\r\n",
      "Inversions                         0                    0\r\n",
      "\r\n",
      "Insertions                       330                  338\r\n",
      "InsertionSum                  231868               204819\r\n",
      "InsertionAvg                702.6303             605.9734\r\n",
      "\r\n",
      "TandemIns                          0                    1\r\n",
      "TandemInsSum                       0                   90\r\n",
      "TandemInsAvg                  0.0000              90.0000\r\n",
      "\r\n",
      "[SNPs]\r\n",
      "TotalSNPs                     303282               303282\r\n",
      "CT                   58698(19.3543%)      51971(17.1362%)\r\n",
      "CA                    11997(3.9557%)       10989(3.6234%)\r\n",
      "CG                    12824(4.2284%)       13781(4.5440%)\r\n",
      "GA                   51947(17.1283%)      47436(15.6409%)\r\n",
      "GT                    13351(4.4022%)       11875(3.9155%)\r\n",
      "GC                    13781(4.5440%)       12824(4.2284%)\r\n",
      "AG                   47436(15.6409%)      51947(17.1283%)\r\n",
      "AC                    10989(3.6234%)       11997(3.9557%)\r\n",
      "AT                     9423(3.1070%)        8990(2.9642%)\r\n",
      "TA                     8990(2.9642%)        9423(3.1070%)\r\n",
      "TC                   51971(17.1362%)      58698(19.3543%)\r\n",
      "TG                    11875(3.9155%)       13351(4.4022%)\r\n",
      "\r\n",
      "TotalGSNPs                      2100                 2100\r\n",
      "AT                       12(0.5714%)          14(0.6667%)\r\n",
      "AC                       31(1.4762%)          19(0.9048%)\r\n",
      "AG                     431(20.5238%)        443(21.0952%)\r\n",
      "TA                       14(0.6667%)          12(0.5714%)\r\n",
      "TC                     484(23.0476%)        524(24.9524%)\r\n",
      "TG                       41(1.9524%)          31(1.4762%)\r\n",
      "CG                       33(1.5714%)          37(1.7619%)\r\n",
      "CA                       19(0.9048%)          31(1.4762%)\r\n",
      "CT                     524(24.9524%)        484(23.0476%)\r\n",
      "GT                       31(1.4762%)          41(1.9524%)\r\n",
      "GA                     443(21.0952%)        431(20.5238%)\r\n",
      "GC                       37(1.7619%)          33(1.5714%)\r\n",
      "\r\n",
      "TotalIndels                   267970               267970\r\n",
      "C.                   50606(18.8849%)       18811(7.0198%)\r\n",
      "G.                   48031(17.9240%)       18375(6.8571%)\r\n",
      "A.                   46970(17.5281%)       19029(7.1012%)\r\n",
      "T.                   45847(17.1090%)       20301(7.5758%)\r\n",
      ".G                    18375(6.8571%)      48031(17.9240%)\r\n",
      ".C                    18811(7.0198%)      50606(18.8849%)\r\n",
      ".A                    19029(7.1012%)      46970(17.5281%)\r\n",
      ".T                    20301(7.5758%)      45847(17.1090%)\r\n",
      "\r\n",
      "TotalGIndels                    2283                 2283\r\n",
      "A.                     407(17.8274%)         201(8.8042%)\r\n",
      "T.                     382(16.7324%)         211(9.2422%)\r\n",
      "C.                     366(16.0315%)         175(7.6654%)\r\n",
      "G.                     359(15.7249%)         182(7.9720%)\r\n",
      ".A                      201(8.8042%)        407(17.8274%)\r\n",
      ".T                      211(9.2422%)        382(16.7324%)\r\n",
      ".G                      182(7.9720%)        359(15.7249%)\r\n",
      ".C                      175(7.6654%)        366(16.0315%)\r\n"
     ]
    }
   ],
   "source": [
    "!cat bs_assembly_miniasm.report"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "When we mapped all our reads to the reference genome in session 2, we saw that numerous reads cover each base in the reference. \n",
    "This information can be used to ammend errors that happened during sequencing and basecalling by aligning all the sequences to our assembly. The general idea is to create an alignment pile of all reads from which we can infer the most frequent base at each position in the assembly.\n",
    "They are differents polishing tools for corecting raw contigs, using as an entry the nanopore reads or illumina reads to perform corrections.\n",
    "The popular polishing tool is racon (https://github.com/isovic/racon); medaka (https://github.com/nanoporetech/medaka); and nanopolish (https://github.com/jts/nanopolish)\n",
    "\n",
    "For a fixed set of parameters, `racon` will ultimately hit the maximal accuracy value after a couple of iterations. To increase accuracy even further, we need to use a different algorithm or even additional information. We have two options, use `medaka` from Oxford Nanopore Technologies which uses deep neural networks trained on `racon` output using only basecalled reads, or use `nanopolish` which uses the raw signal data to increase the accuracy with hidden Markov models. First we will try `medaka` which is much faster than `nanopolish`, and afterwards we will showcase how to use `nanopolish`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# POLISHING WITH RACON"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The consensus module `racon` was developed atop `minimap` and `miniasm` assembly pipeline as the consensus module for third generation sequencing data. The core engine of `racon` is a partial order alignment library called `spoa`.  \n",
    "(see https://github.com/isovic/racon)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "`racon` was designed to iteratively increase the accuracy of a target sequence by first using a mapper (`minimap2`) to map/align all the reads to the target. Afterwards, it filters out low quality overlaps, slices the target sequence into windows of 500 bp, drops read parts that do not pass a quality threshold and construct a multiple sequence alignment from each window. After calling consensus on each window, the final sequence is obtained by concatenating all window consensuses.  \n",
    "\n",
    "(see https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5411768/)\n",
    "\n",
    "We are going to use racon to do an initial correction. The medaka documentation advises to do four rounds with racon before polishing with medaka since medaka has been trained with racon polished assemblies. \n",
    "\n",
    "(see https://denbi-nanopore-training-course.readthedocs.io/en/latest/polishing/index.html)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<img src=\"https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5411768/bin/737f03.jpg\"/>"
      ],
      "text/plain": [
       "<IPython.core.display.Image object>"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from IPython.display import Image\n",
    "Image(url =\"https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5411768/bin/737f03.jpg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!racon"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let us run first `minimap2` to find the positions where our reads map to the assembly and then we will use `racon` to increase the accuracy. Eventually running `dnadiff` will give us the details about the differences against the reference genome.  \n",
    "**Do 4 rounds of Racon**."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Round 1**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[M::mm_idx_gen::0.162*1.01] collected minimizers\n",
      "[M::mm_idx_gen::0.184*1.34] sorted minimizers\n",
      "[M::main::0.184*1.34] loaded/built the index for 1 target sequence(s)\n",
      "[M::mm_mapopt_update::0.192*1.33] mid_occ = 10\n",
      "[M::mm_idx_stat] kmer size: 15; skip: 10; is_hpc: 0; #seq: 1\n",
      "[M::mm_idx_stat::0.199*1.32] distinct minimizers: 728417 (99.14% are singletons); average occurrences: 1.009; average spacing: 5.347; total length: 3931083\n",
      "[M::worker_pipeline::6.801*2.59] mapped 26277 sequences\n",
      "[M::main] Version: 2.24-r1122\n",
      "[M::main] CMD: minimap2 -t 4 -x map-ont bs_assembly_miniasm.fasta data/bacillus_subtilis/bs_reads.fastq.gz\n",
      "[M::main] Real time: 6.823 sec; CPU: 17.615 sec; Peak RSS: 0.233 GB\n",
      "[racon::Polisher::initialize] loaded target sequences 0.035785 s\n",
      "[racon::Polisher::initialize] loaded sequences 4.801240 s\n",
      "[racon::Polisher::initialize] loaded overlaps 0.048462 s\n",
      "[racon::Polisher::initialize] aligning overlaps [====================] 104.317027 s\n",
      "[racon::Polisher::initialize] transformed data into windows 0.358776 s\n",
      "[racon::Polisher::polish] generating consensus [====================] 158.007180 s\n",
      "[racon::Polisher::] total = 267.710516 s\n"
     ]
    }
   ],
   "source": [
    "!minimap2 \\\n",
    "    -t 4 \\\n",
    "    -x map-ont \\\n",
    "    bs_assembly_miniasm.fasta \\\n",
    "    data/bacillus_subtilis/bs_reads.fastq.gz > bs_assembly_miniasm.paf\n",
    "\n",
    "!racon \\\n",
    "    -t 4 \\\n",
    "    data/bacillus_subtilis/bs_reads.fastq.gz \\\n",
    "    bs_assembly_miniasm.paf \\\n",
    "    bs_assembly_miniasm.fasta > bs_assembly_miniasm_r1.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "!dnadiff \\\n",
    "    -p bs_assembly_miniasm_r1 \\\n",
    "    data/bacillus_subtilis/bs_ref.fasta \\\n",
    "    bs_assembly_miniasm_r1.fasta 2> err"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/guyot/Piracicaba_ESALQ_2023/Session_4/data/bacillus_subtilis/bs_ref.fasta /Users/guyot/Piracicaba_ESALQ_2023/Session_4/bs_assembly_miniasm_r1.fasta\r\n",
      "NUCMER\r\n",
      "\r\n",
      "                               [REF]                [QRY]\r\n",
      "[Sequences]\r\n",
      "TotalSeqs                          1                    1\r\n",
      "AlignedSeqs              1(100.0000%)          1(100.0000%)\r\n",
      "UnalignedSeqs             0(0.0000%)           0(0.0000%)\r\n",
      "\r\n",
      "[Bases]\r\n",
      "TotalBases                   4045677              4043714\r\n",
      "AlignedBases       4045093(99.9856%)    4039207(99.8885%)\r\n",
      "UnalignedBases          584(0.0144%)        4507(0.1115%)\r\n",
      "\r\n",
      "[Alignments]\r\n",
      "1-to-1                             6                    6\r\n",
      "TotalLength                  4060621              4039311\r\n",
      "AvgLength                676770.1667          673218.5000\r\n",
      "AvgIdentity                  98.7446              98.7446\r\n",
      "\r\n",
      "M-to-M                             6                    6\r\n",
      "TotalLength                  4060621              4039311\r\n",
      "AvgLength                676770.1667          673218.5000\r\n",
      "AvgIdentity                  98.7446              98.7446\r\n",
      "\r\n",
      "[Feature Estimates]\r\n",
      "Breakpoints                       10                   11\r\n",
      "Relocations                        1                    1\r\n",
      "Translocations                     0                    0\r\n",
      "Inversions                         0                    0\r\n",
      "\r\n",
      "Insertions                         3                    4\r\n",
      "InsertionSum                     687                 4620\r\n",
      "InsertionAvg                229.0000            1155.0000\r\n",
      "\r\n",
      "TandemIns                          1                    0\r\n",
      "TandemInsSum                     103                    0\r\n",
      "TandemInsAvg                103.0000               0.0000\r\n",
      "\r\n",
      "[SNPs]\r\n",
      "TotalSNPs                      13600                13600\r\n",
      "AC                      175(1.2868%)         284(2.0882%)\r\n",
      "AG                    1876(13.7941%)       4578(33.6618%)\r\n",
      "AT                      289(2.1250%)         198(1.4559%)\r\n",
      "GT                      444(3.2647%)         159(1.1691%)\r\n",
      "GA                    4578(33.6618%)       1876(13.7941%)\r\n",
      "GC                      326(2.3971%)         272(2.0000%)\r\n",
      "TC                     1254(9.2206%)       3745(27.5368%)\r\n",
      "TG                      159(1.1691%)         444(3.2647%)\r\n",
      "TA                      198(1.4559%)         289(2.1250%)\r\n",
      "CA                      284(2.0882%)         175(1.2868%)\r\n",
      "CT                    3745(27.5368%)        1254(9.2206%)\r\n",
      "CG                      272(2.0000%)         326(2.3971%)\r\n",
      "\r\n",
      "TotalGSNPs                      5560                 5560\r\n",
      "AT                       49(0.8813%)          20(0.3597%)\r\n",
      "AG                     955(17.1763%)       2053(36.9245%)\r\n",
      "AC                       25(0.4496%)          47(0.8453%)\r\n",
      "GC                       55(0.9892%)          50(0.8993%)\r\n",
      "GT                       81(1.4568%)          24(0.4317%)\r\n",
      "GA                    2053(36.9245%)        955(17.1763%)\r\n",
      "TC                     643(11.5647%)       1558(28.0216%)\r\n",
      "TA                       20(0.3597%)          49(0.8813%)\r\n",
      "TG                       24(0.4317%)          81(1.4568%)\r\n",
      "CA                       47(0.8453%)          25(0.4496%)\r\n",
      "CG                       50(0.8993%)          55(0.9892%)\r\n",
      "CT                    1558(28.0216%)        643(11.5647%)\r\n",
      "\r\n",
      "TotalIndels                    36622                36622\r\n",
      "A.                    4925(13.4482%)        2666(7.2798%)\r\n",
      "G.                   10389(28.3682%)         931(2.5422%)\r\n",
      "T.                     3611(9.8602%)        3282(8.9618%)\r\n",
      "C.                    9892(27.0111%)         926(2.5285%)\r\n",
      ".T                     3282(8.9618%)        3611(9.8602%)\r\n",
      ".G                      931(2.5422%)      10389(28.3682%)\r\n",
      ".A                     2666(7.2798%)       4925(13.4482%)\r\n",
      ".C                      926(2.5285%)       9892(27.0111%)\r\n",
      "\r\n",
      "TotalGIndels                   16537                16537\r\n",
      "A.                    2244(13.5696%)        1013(6.1257%)\r\n",
      "G.                    5340(32.2912%)         262(1.5843%)\r\n",
      "T.                     1436(8.6836%)        1137(6.8755%)\r\n",
      "C.                    4871(29.4552%)         234(1.4150%)\r\n",
      ".A                     1013(6.1257%)       2244(13.5696%)\r\n",
      ".G                      262(1.5843%)       5340(32.2912%)\r\n",
      ".T                     1137(6.8755%)        1436(8.6836%)\r\n",
      ".C                      234(1.4150%)       4871(29.4552%)\r\n"
     ]
    }
   ],
   "source": [
    "!cat bs_assembly_miniasm_r1.report"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "With only one `racon` iteration the accuracy increased from $85.28\\%$ to $98.76\\%$. In addition, the number of break points decreased drastically, the assembly length almost matches the actual reference length, and almost no unalignable bases are left."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We will run a few more iterations to see how far we can increase the accuracy.  \n",
    "Do several iterations of `Racon`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Round 2**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[M::mm_idx_gen::0.180*0.99] collected minimizers\n",
      "[M::mm_idx_gen::0.205*1.33] sorted minimizers\n",
      "[M::main::0.205*1.33] loaded/built the index for 1 target sequence(s)\n",
      "[M::mm_mapopt_update::0.218*1.31] mid_occ = 11\n",
      "[M::mm_idx_stat] kmer size: 15; skip: 10; is_hpc: 0; #seq: 1\n",
      "[M::mm_idx_stat::0.231*1.28] distinct minimizers: 737882 (98.44% are singletons); average occurrences: 1.026; average spacing: 5.343; total length: 4043714\n",
      "[M::worker_pipeline::9.321*2.64] mapped 26277 sequences\n",
      "[M::main] Version: 2.24-r1122\n",
      "[M::main] CMD: minimap2 -t 4 -x map-ont bs_assembly_miniasm_r1.fasta data/bacillus_subtilis/bs_reads.fastq.gz\n",
      "[M::main] Real time: 9.349 sec; CPU: 24.617 sec; Peak RSS: 0.233 GB\n",
      "[racon::Polisher::initialize] loaded target sequences 0.054751 s\n",
      "[racon::Polisher::initialize] loaded sequences 5.163677 s\n",
      "[racon::Polisher::initialize] loaded overlaps 0.053287 s\n",
      "[racon::Polisher::initialize] aligning overlaps [====================] 115.560905 s\n",
      "[racon::Polisher::initialize] transformed data into windows 0.485339 s\n",
      "[racon::Polisher::polish] generating consensus [====================] 356.063354 s\n",
      "[racon::Polisher::] total = 477.598795 s\n"
     ]
    }
   ],
   "source": [
    "!minimap2 \\\n",
    "    -t 4 \\\n",
    "    -x map-ont \\\n",
    "    bs_assembly_miniasm_r1.fasta \\\n",
    "    data/bacillus_subtilis/bs_reads.fastq.gz > bs_assembly_miniasm_r1.paf\n",
    "\n",
    "!racon \\\n",
    "    -t 4 \\\n",
    "    data/bacillus_subtilis/bs_reads.fastq.gz \\\n",
    "   bs_assembly_miniasm_r1.paf \\\n",
    "    bs_assembly_miniasm_r1.fasta > bs_assembly_miniasm_r2.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/guyot/Piracicaba_ESALQ_2023/Session_4/data/bacillus_subtilis/bs_ref.fasta /Users/guyot/Piracicaba_ESALQ_2023/Session_4/bs_assembly_miniasm_r2.fasta\r\n",
      "NUCMER\r\n",
      "\r\n",
      "                               [REF]                [QRY]\r\n",
      "[Sequences]\r\n",
      "TotalSeqs                          1                    1\r\n",
      "AlignedSeqs              1(100.0000%)          1(100.0000%)\r\n",
      "UnalignedSeqs             0(0.0000%)           0(0.0000%)\r\n",
      "\r\n",
      "[Bases]\r\n",
      "TotalBases                   4045677              4047139\r\n",
      "AlignedBases       4045677(100.0000%)    4043151(99.9015%)\r\n",
      "UnalignedBases            0(0.0000%)        3988(0.0985%)\r\n",
      "\r\n",
      "[Alignments]\r\n",
      "1-to-1                             3                    3\r\n",
      "TotalLength                  4061092              4043254\r\n",
      "AvgLength               1353697.3333         1347751.3333\r\n",
      "AvgIdentity                  99.0268              99.0268\r\n",
      "\r\n",
      "M-to-M                             3                    3\r\n",
      "TotalLength                  4061092              4043254\r\n",
      "AvgLength               1353697.3333         1347751.3333\r\n",
      "AvgIdentity                  99.0268              99.0268\r\n",
      "\r\n",
      "[Feature Estimates]\r\n",
      "Breakpoints                        4                    5\r\n",
      "Relocations                        1                    1\r\n",
      "Translocations                     0                    0\r\n",
      "Inversions                         0                    0\r\n",
      "\r\n",
      "Insertions                         1                    1\r\n",
      "InsertionSum                     102                 3988\r\n",
      "InsertionAvg                102.0000            3988.0000\r\n",
      "\r\n",
      "TandemIns                          1                    0\r\n",
      "TandemInsSum                     102                    0\r\n",
      "TandemInsAvg                102.0000               0.0000\r\n",
      "\r\n",
      "[SNPs]\r\n",
      "TotalSNPs                      10772                10772\r\n",
      "TA                      123(1.1418%)         170(1.5782%)\r\n",
      "TC                      937(8.6985%)       3011(27.9521%)\r\n",
      "TG                      104(0.9655%)         334(3.1006%)\r\n",
      "GC                      203(1.8845%)         180(1.6710%)\r\n",
      "GA                    3882(36.0379%)       1536(14.2592%)\r\n",
      "GT                      334(3.1006%)         104(0.9655%)\r\n",
      "CT                    3011(27.9521%)         937(8.6985%)\r\n",
      "CG                      180(1.6710%)         203(1.8845%)\r\n",
      "CA                      184(1.7081%)         108(1.0026%)\r\n",
      "AT                      170(1.5782%)         123(1.1418%)\r\n",
      "AC                      108(1.0026%)         184(1.7081%)\r\n",
      "AG                    1536(14.2592%)       3882(36.0379%)\r\n",
      "\r\n",
      "TotalGSNPs                      5077                 5077\r\n",
      "TA                       18(0.3545%)          29(0.5712%)\r\n",
      "TG                       20(0.3939%)          67(1.3197%)\r\n",
      "TC                     557(10.9710%)       1371(27.0041%)\r\n",
      "CG                       41(0.8076%)          47(0.9257%)\r\n",
      "CA                       49(0.9651%)          22(0.4333%)\r\n",
      "CT                    1371(27.0041%)        557(10.9710%)\r\n",
      "GC                       47(0.9257%)          41(0.8076%)\r\n",
      "GA                    1968(38.7630%)        888(17.4906%)\r\n",
      "GT                       67(1.3197%)          20(0.3939%)\r\n",
      "AG                     888(17.4906%)       1968(38.7630%)\r\n",
      "AC                       22(0.4333%)          49(0.9651%)\r\n",
      "AT                       29(0.5712%)          18(0.3545%)\r\n",
      "\r\n",
      "TotalIndels                    27784                27784\r\n",
      "T.                     2176(7.8318%)        2331(8.3897%)\r\n",
      "G.                    8902(32.0400%)         512(1.8428%)\r\n",
      "C.                    8397(30.2224%)         489(1.7600%)\r\n",
      "A.                    3189(11.4778%)        1788(6.4354%)\r\n",
      ".T                     2331(8.3897%)        2176(7.8318%)\r\n",
      ".C                      489(1.7600%)       8397(30.2224%)\r\n",
      ".G                      512(1.8428%)       8902(32.0400%)\r\n",
      ".A                     1788(6.4354%)       3189(11.4778%)\r\n",
      "\r\n",
      "TotalGIndels                   15103                15103\r\n",
      "T.                     1137(7.5283%)         919(6.0849%)\r\n",
      "C.                    4783(31.6692%)         162(1.0726%)\r\n",
      "G.                    5319(35.2182%)         193(1.2779%)\r\n",
      "A.                    1755(11.6202%)         835(5.5287%)\r\n",
      ".A                      835(5.5287%)       1755(11.6202%)\r\n",
      ".G                      193(1.2779%)       5319(35.2182%)\r\n",
      ".C                      162(1.0726%)       4783(31.6692%)\r\n",
      ".T                      919(6.0849%)        1137(7.5283%)\r\n"
     ]
    }
   ],
   "source": [
    "!dnadiff \\\n",
    "    -p bs_assembly_miniasm_r2 \\\n",
    "    data/bacillus_subtilis/bs_ref.fasta \\\n",
    "    bs_assembly_miniasm_r2.fasta 2> err\n",
    "\n",
    "!cat bs_assembly_miniasm_r2.report"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Round 3**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[M::mm_idx_gen::0.640*0.71] collected minimizers\n",
      "[M::mm_idx_gen::0.675*0.84] sorted minimizers\n",
      "[M::main::0.675*0.84] loaded/built the index for 1 target sequence(s)\n",
      "[M::mm_mapopt_update::0.694*0.84] mid_occ = 11\n",
      "[M::mm_idx_stat] kmer size: 15; skip: 10; is_hpc: 0; #seq: 1\n",
      "[M::mm_idx_stat::0.711*0.84] distinct minimizers: 737952 (98.45% are singletons); average occurrences: 1.027; average spacing: 5.342; total length: 4047139\n",
      "[M::worker_pipeline::11.408*2.46] mapped 26277 sequences\n",
      "[M::main] Version: 2.24-r1122\n",
      "[M::main] CMD: minimap2 -t 4 -x map-ont bs_assembly_miniasm_r2.fasta data/bacillus_subtilis/bs_reads.fastq.gz\n",
      "[M::main] Real time: 11.438 sec; CPU: 28.068 sec; Peak RSS: 0.233 GB\n",
      "[racon::Polisher::initialize] loaded target sequences 0.047200 s\n",
      "[racon::Polisher::initialize] loaded sequences 5.278702 s\n",
      "[racon::Polisher::initialize] loaded overlaps 0.043658 s\n",
      "[racon::Polisher::initialize] aligning overlaps [====================] 105.056373 s\n",
      "[racon::Polisher::initialize] transformed data into windows 0.379885 s\n",
      "[racon::Polisher::polish] generating consensus [====================] 340.591849 s\n",
      "[racon::Polisher::] total = 451.666076 s\n"
     ]
    }
   ],
   "source": [
    "!minimap2 \\\n",
    "    -t 4 \\\n",
    "    -x map-ont \\\n",
    "    bs_assembly_miniasm_r2.fasta \\\n",
    "    data/bacillus_subtilis/bs_reads.fastq.gz > bs_assembly_miniasm_r2.paf\n",
    "\n",
    "!racon \\\n",
    "    -t 4 \\\n",
    "    data/bacillus_subtilis/bs_reads.fastq.gz \\\n",
    "   bs_assembly_miniasm_r2.paf \\\n",
    "    bs_assembly_miniasm_r2.fasta > bs_assembly_miniasm_r3.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cat: bs_assembly_miniasm_r3.report: No such file or directory\r\n"
     ]
    }
   ],
   "source": [
    "!dnadiff \\\n",
    "    -p bs_assembly_miniasm_r3 \\\n",
    "    data/bacillus_subtilis/bs_ref.fasta \\\n",
    "    bs_assembly_miniasm_r3.fasta 2> err\n",
    "\n",
    "!cat bs_assembly_miniasm_r3.report"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/guyot/Piracicaba_ESALQ_2023/Session_4/data/bacillus_subtilis/bs_ref.fasta /Users/guyot/Piracicaba_ESALQ_2023/Session_4/bs_assembly_miniasm_r3.fasta\r\n",
      "NUCMER\r\n",
      "\r\n",
      "                               [REF]                [QRY]\r\n",
      "[Sequences]\r\n",
      "TotalSeqs                          1                    1\r\n",
      "AlignedSeqs              1(100.0000%)          1(100.0000%)\r\n",
      "UnalignedSeqs             0(0.0000%)           0(0.0000%)\r\n",
      "\r\n",
      "[Bases]\r\n",
      "TotalBases                   4045677              4047661\r\n",
      "AlignedBases       4045677(100.0000%)    4043673(99.9015%)\r\n",
      "UnalignedBases            0(0.0000%)        3988(0.0985%)\r\n",
      "\r\n",
      "[Alignments]\r\n",
      "1-to-1                             3                    3\r\n",
      "TotalLength                  4061092              4043776\r\n",
      "AvgLength               1353697.3333         1347925.3333\r\n",
      "AvgIdentity                  99.0676              99.0676\r\n",
      "\r\n",
      "M-to-M                             3                    3\r\n",
      "TotalLength                  4061092              4043776\r\n",
      "AvgLength               1353697.3333         1347925.3333\r\n",
      "AvgIdentity                  99.0676              99.0676\r\n",
      "\r\n",
      "[Feature Estimates]\r\n",
      "Breakpoints                        4                    5\r\n",
      "Relocations                        1                    1\r\n",
      "Translocations                     0                    0\r\n",
      "Inversions                         0                    0\r\n",
      "\r\n",
      "Insertions                         1                    1\r\n",
      "InsertionSum                     102                 3988\r\n",
      "InsertionAvg                102.0000            3988.0000\r\n",
      "\r\n",
      "TandemIns                          1                    0\r\n",
      "TandemInsSum                     102                    0\r\n",
      "TandemInsAvg                102.0000               0.0000\r\n",
      "\r\n",
      "[SNPs]\r\n",
      "TotalSNPs                      10319                10319\r\n",
      "AC                       89(0.8625%)         151(1.4633%)\r\n",
      "AT                      136(1.3180%)         111(1.0757%)\r\n",
      "AG                    1464(14.1874%)       3816(36.9803%)\r\n",
      "GT                      298(2.8879%)          76(0.7365%)\r\n",
      "GC                      190(1.8413%)         160(1.5505%)\r\n",
      "GA                    3816(36.9803%)       1464(14.1874%)\r\n",
      "CT                    2904(28.1423%)         924(8.9544%)\r\n",
      "CA                      151(1.4633%)          89(0.8625%)\r\n",
      "CG                      160(1.5505%)         190(1.8413%)\r\n",
      "TC                      924(8.9544%)       2904(28.1423%)\r\n",
      "TA                      111(1.0757%)         136(1.3180%)\r\n",
      "TG                       76(0.7365%)         298(2.8879%)\r\n",
      "\r\n",
      "TotalGSNPs                      5036                 5036\r\n",
      "GC                       48(0.9531%)          41(0.8141%)\r\n",
      "GT                       62(1.2311%)          17(0.3376%)\r\n",
      "GA                    1963(38.9793%)        873(17.3352%)\r\n",
      "AG                     873(17.3352%)       1963(38.9793%)\r\n",
      "AC                       19(0.3773%)          40(0.7943%)\r\n",
      "AT                       34(0.6751%)          13(0.2581%)\r\n",
      "TG                       17(0.3376%)          62(1.2311%)\r\n",
      "TA                       13(0.2581%)          34(0.6751%)\r\n",
      "TC                     552(10.9611%)       1374(27.2836%)\r\n",
      "CG                       41(0.8141%)          48(0.9531%)\r\n",
      "CA                       40(0.7943%)          19(0.3773%)\r\n",
      "CT                    1374(27.2836%)        552(10.9611%)\r\n",
      "\r\n",
      "TotalIndels                    26725                26725\r\n",
      "A.                    3010(11.2629%)        1717(6.4247%)\r\n",
      "G.                    8699(32.5500%)         477(1.7848%)\r\n",
      "C.                    8167(30.5594%)         426(1.5940%)\r\n",
      "T.                     2002(7.4911%)        2227(8.3330%)\r\n",
      ".G                      477(1.7848%)       8699(32.5500%)\r\n",
      ".A                     1717(6.4247%)       3010(11.2629%)\r\n",
      ".T                     2227(8.3330%)        2002(7.4911%)\r\n",
      ".C                      426(1.5940%)       8167(30.5594%)\r\n",
      "\r\n",
      "TotalGIndels                   14999                14999\r\n",
      "G.                    5353(35.6890%)         203(1.3534%)\r\n",
      "A.                    1726(11.5074%)         791(5.2737%)\r\n",
      "T.                     1101(7.3405%)         909(6.0604%)\r\n",
      "C.                    4770(31.8021%)         146(0.9734%)\r\n",
      ".G                      203(1.3534%)       5353(35.6890%)\r\n",
      ".A                      791(5.2737%)       1726(11.5074%)\r\n",
      ".T                      909(6.0604%)        1101(7.3405%)\r\n",
      ".C                      146(0.9734%)       4770(31.8021%)\r\n"
     ]
    }
   ],
   "source": [
    "!cat bs_assembly_miniasm_3.report"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Round 4**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[M::mm_idx_gen::0.313*0.92] collected minimizers\n",
      "[M::mm_idx_gen::0.337*1.14] sorted minimizers\n",
      "[M::main::0.337*1.14] loaded/built the index for 1 target sequence(s)\n",
      "[M::mm_mapopt_update::0.353*1.13] mid_occ = 11\n",
      "[M::mm_idx_stat] kmer size: 15; skip: 10; is_hpc: 0; #seq: 1\n",
      "[M::mm_idx_stat::0.368*1.12] distinct minimizers: 737911 (98.46% are singletons); average occurrences: 1.027; average spacing: 5.342; total length: 4047661\n",
      "[M::worker_pipeline::34.581*1.29] mapped 26277 sequences\n",
      "[M::main] Version: 2.24-r1122\n",
      "[M::main] CMD: minimap2 -t 4 -x map-ont bs_assembly_miniasm_r3.fasta data/bacillus_subtilis/bs_reads.fastq.gz\n",
      "[M::main] Real time: 34.616 sec; CPU: 44.579 sec; Peak RSS: 0.233 GB\n",
      "[racon::Polisher::initialize] loaded target sequences 0.076103 s\n",
      "[racon::Polisher::initialize] loaded sequences 15.133805 s\n",
      "[racon::Polisher::initialize] loaded overlaps 0.125352 s\n",
      "[racon::Polisher::initialize] aligning overlaps [====================] 178.464364 s\n",
      "[racon::Polisher::initialize] transformed data into windows 0.308335 s\n",
      "[racon::Polisher::polish] generating consensus [====================] 184.340539 s\n",
      "[racon::Polisher::] total = 378.570601 s\n"
     ]
    }
   ],
   "source": [
    "!minimap2 \\\n",
    "    -t 4 \\\n",
    "    -x map-ont \\\n",
    "    bs_assembly_miniasm_r3.fasta \\\n",
    "    data/bacillus_subtilis/bs_reads.fastq.gz > bs_assembly_miniasm_r3.paf\n",
    "\n",
    "!racon \\\n",
    "    -t 4 \\\n",
    "    data/bacillus_subtilis/bs_reads.fastq.gz \\\n",
    "   bs_assembly_miniasm_r3.paf \\\n",
    "    bs_assembly_miniasm_r3.fasta > bs_assembly_miniasm_r4.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cat: bs_assembly_miniasm_r4.report: No such file or directory\r\n"
     ]
    }
   ],
   "source": [
    "!dnadiff \\\n",
    "    -p bs_assembly_miniasm_r4 \\\n",
    "    data/bacillus_subtilis/bs_ref.fasta \\\n",
    "    bs_assembly_miniasm_r4.fasta 2> err\n",
    "\n",
    "!cat bs_assembly_miniasm_r4.report"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/guyot/Piracicaba_ESALQ_2023/Session_4/data/bacillus_subtilis/bs_ref.fasta /Users/guyot/Piracicaba_ESALQ_2023/Session_4/bs_assembly_miniasm_r4.fasta\r\n",
      "NUCMER\r\n",
      "\r\n",
      "                               [REF]                [QRY]\r\n",
      "[Sequences]\r\n",
      "TotalSeqs                          1                    1\r\n",
      "AlignedSeqs              1(100.0000%)          1(100.0000%)\r\n",
      "UnalignedSeqs             0(0.0000%)           0(0.0000%)\r\n",
      "\r\n",
      "[Bases]\r\n",
      "TotalBases                   4045677              4047966\r\n",
      "AlignedBases       4045677(100.0000%)    4043978(99.9015%)\r\n",
      "UnalignedBases            0(0.0000%)        3988(0.0985%)\r\n",
      "\r\n",
      "[Alignments]\r\n",
      "1-to-1                             3                    3\r\n",
      "TotalLength                  4061092              4044080\r\n",
      "AvgLength               1353697.3333         1348026.6667\r\n",
      "AvgIdentity                  99.0749              99.0749\r\n",
      "\r\n",
      "M-to-M                             3                    3\r\n",
      "TotalLength                  4061092              4044080\r\n",
      "AvgLength               1353697.3333         1348026.6667\r\n",
      "AvgIdentity                  99.0749              99.0749\r\n",
      "\r\n",
      "[Feature Estimates]\r\n",
      "Breakpoints                        4                    5\r\n",
      "Relocations                        1                    1\r\n",
      "Translocations                     0                    0\r\n",
      "Inversions                         0                    0\r\n",
      "\r\n",
      "Insertions                         1                    1\r\n",
      "InsertionSum                     101                 3988\r\n",
      "InsertionAvg                101.0000            3988.0000\r\n",
      "\r\n",
      "TandemIns                          1                    0\r\n",
      "TandemInsSum                     101                    0\r\n",
      "TandemInsAvg                101.0000               0.0000\r\n",
      "\r\n",
      "[SNPs]\r\n",
      "TotalSNPs                      10208                10208\r\n",
      "GT                      294(2.8801%)          75(0.7347%)\r\n",
      "GA                    3780(37.0298%)       1482(14.5180%)\r\n",
      "GC                      186(1.8221%)         148(1.4498%)\r\n",
      "TC                      896(8.7774%)       2859(28.0074%)\r\n",
      "TA                       94(0.9208%)         140(1.3715%)\r\n",
      "TG                       75(0.7347%)         294(2.8801%)\r\n",
      "AC                       95(0.9306%)         159(1.5576%)\r\n",
      "AG                    1482(14.5180%)       3780(37.0298%)\r\n",
      "AT                      140(1.3715%)          94(0.9208%)\r\n",
      "CT                    2859(28.0074%)         896(8.7774%)\r\n",
      "CG                      148(1.4498%)         186(1.8221%)\r\n",
      "CA                      159(1.5576%)          95(0.9306%)\r\n",
      "\r\n",
      "TotalGSNPs                      5017                 5017\r\n",
      "CG                       42(0.8372%)          49(0.9767%)\r\n",
      "CT                    1342(26.7491%)        548(10.9229%)\r\n",
      "CA                       40(0.7973%)          18(0.3588%)\r\n",
      "AT                       34(0.6777%)          10(0.1993%)\r\n",
      "AG                     886(17.6600%)       1955(38.9675%)\r\n",
      "AC                       18(0.3588%)          40(0.7973%)\r\n",
      "GA                    1955(38.9675%)        886(17.6600%)\r\n",
      "GT                       76(1.5148%)          17(0.3388%)\r\n",
      "GC                       49(0.9767%)          42(0.8372%)\r\n",
      "TA                       10(0.1993%)          34(0.6777%)\r\n",
      "TG                       17(0.3388%)          76(1.5148%)\r\n",
      "TC                     548(10.9229%)       1342(26.7491%)\r\n",
      "\r\n",
      "TotalIndels                    26534                26534\r\n",
      "G.                    8650(32.5997%)         473(1.7826%)\r\n",
      "T.                     1910(7.1983%)        2259(8.5136%)\r\n",
      "A.                    2952(11.1253%)        1706(6.4295%)\r\n",
      "C.                    8120(30.6022%)         464(1.7487%)\r\n",
      ".C                      464(1.7487%)       8120(30.6022%)\r\n",
      ".T                     2259(8.5136%)        1910(7.1983%)\r\n",
      ".G                      473(1.7826%)       8650(32.5997%)\r\n",
      ".A                     1706(6.4295%)       2952(11.1253%)\r\n",
      "\r\n",
      "TotalGIndels                   15080                15080\r\n",
      "C.                    4811(31.9032%)         168(1.1141%)\r\n",
      "A.                    1726(11.4456%)         788(5.2255%)\r\n",
      "G.                    5401(35.8156%)         188(1.2467%)\r\n",
      "T.                     1075(7.1286%)         923(6.1207%)\r\n",
      ".C                      168(1.1141%)       4811(31.9032%)\r\n",
      ".A                      788(5.2255%)       1726(11.4456%)\r\n",
      ".T                      923(6.1207%)        1075(7.1286%)\r\n",
      ".G                      188(1.2467%)       5401(35.8156%)\r\n"
     ]
    }
   ],
   "source": [
    "!cat bs_assembly_miniasm_4.report"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# POLISHING WITH MEDAKA"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "`medaka` is a tool to create consensus sequences and variant calls from nanopore sequencing data. This task is performed using neural networks applied a pileup of individual sequencing reads against a draft assembly. It provides state-of-the-art results outperforming sequence-graph based methods and signal-based methods, whilst also being faster.\n",
    "\n",
    "(see https://github.com/nanoporetech/medaka)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Medaka is not compatible with the curso_4 env. **A new environment is necessary**. In a new terminal type : "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "!conda create -y --name medaka python=3.6 "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "!conda install -y -n medaka medaka jupyter"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "!conda activate medaka"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "!jupyter notebook &"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!medaka_consensus -h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "!medaka_consensus \\\n",
    "    -i data/bacillus_subtilis/bs_reads.fastq.gz \\\n",
    "    -d bs_assembly_miniasm_r4.fasta \\\n",
    "    -m r10_min_high_g340 \\\n",
    "    -t 4 \\\n",
    "    -o bs_miniasm_medaka \\\n",
    "    -f"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!ls bs_miniasm_medaka"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!dnadiff \\\n",
    "    -p bs_miniasm_medaka \\\n",
    "    data/bacillus_subtilis/bs_ref.fasta \\\n",
    "    bs_miniasm_medaka/consensus.fasta 2> err\n",
    "\n",
    "!cat bs_miniasm_medaka.report"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let us try running `medaka` directly on the unpolished assembly."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!medaka_consensus \\\n",
    "    -i data/bacillus_subtilis/bs_reads.fastq.gz \\\n",
    "    -d data/bacillus_subtilis/ bs_assembly_miniasm.fasta \\\n",
    "    -m r10_min_high_g340 \\\n",
    "    -t 4 \\\n",
    "    -o bs_raw_miniasm_medaka \\\n",
    "    -f"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!dnadiff \\\n",
    "    -p bs_raw_miniasm_medaka \\\n",
    "    data/bacillus_subtilis/bs_ref.fasta \\\n",
    "    bs_raw_miniasm_medaka/consensus.fasta 2> err\n",
    "\n",
    "!cat bs_raw_miniasm_medaka.report"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# POLISHING WITH NANOPOLISH"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "With the fact that `nanopolish`-ing takes a lot of time, we will not run it.  \n",
    "(see https://github.com/jts/nanopolishan).  \n",
    "Software package for signal-level analysis of Oxford Nanopore sequencing data. Nanopolish can calculate an improved consensus sequence for a draft genome assembly, detect base modifications, call SNPs and indels with respect to a reference genome and more (see Nanopolish modules, below)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!nanopolish --help"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First we need to let `nanopolish` connect the raw data (fast5) and the basecalled data. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# OTHERS TOOLS"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**[Pilon](https://github.com/broadinstitute/pilon)**.  **With Illumina data**.  \n",
    "(see https://denbi-nanopore-training-course.readthedocs.io/en/latest/polishing/pilon/index.html)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Comparative evaluation of Nanopore polishing tools"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "See https://doi.org/10.1038/s41598-021-00178-w "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from IPython.display import Image\n",
    "Image(url = \"https://media.springernature.com/full/springer-static/image/art%3A10.1038%2Fs41598-021-00178-w/MediaObjects/41598_2021_178_Fig1_HTML.png?as=webp\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "______________________________________________________________________________________"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**RESULTS**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Complete the following results table :"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "|Assembly Tools| Raw reads | Corrected reads| Number of SNPs | Polishing tools | R1 SNP| R2 SNP | R3 SNP | R4 SNP |Medaka|\n",
    "|---|---|---|---|---|---|---|---|---|---|\n",
    "| Minimap/Miniasm| x| | 303000| Racon| 13000 | 10000| 10000 | 10000 | 1423 |\n",
    "| Flye| x| | 8400  | Racon |9556 | | | |\n",
    "| Flye| x| | 8400  | Medaka | 1030 | | | |\n",
    "| Flye/Canu| | x| 10839| Racon | 9635 | | | |\n",
    "| Flye/Canu| | x| 10839 |Medaka |1012 | | | | |\n",
    "| Raven| x| | 307018| | | | | |\n",
    "| Raven/Canu| | x| 56741| | | | | | |\n",
    "| Canu| | x| 13081| Racon| 9965 | | | | 1369 |\n",
    "| Canu| | x| 13081| Medaka| 1873 | | | |\n",
    "| Shasta| x| | | | | | | | |\n",
    "| Shasta/Canu| | x| | | | | | | |\n",
    "| | | | | | | | | | |\n",
    "| | | | | |  | | | |"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
